### ISSUE TITLE
values with IOK and pNOK (but not NOK) encoded as floating point, losing precision

### ISSUE THREAD

**Issue by haarg** [rurban/Cpanel-JSON-XS#197]
See the following script:
```perl
use strict;
use warnings;
use Cpanel::JSON::XS;
use Devel::Peek;

my $l = 412345678901234567;
Dump($l);
print encode_json([$l]) . "\n";
my $g = $l + 1.2;
Dump($l);
print encode_json([$l]) . "\n";

__END__
SV = IV(0x128023738) at 0x128023748
  REFCNT = 1
  FLAGS = (IOK,pIOK)
  IV = 412345678901234567
[412345678901234567]
SV = PVNV(0x1280098f0) at 0x128023748
  REFCNT = 1
  FLAGS = (IOK,pIOK,pNOK)
  IV = 412345678901234567
  NV = 4.12345678901235e+17
  PV = 0
[4.12345678901235e+17]
```
Using `$l` as a float upgraded the value to PVNV and set the pNOK flag. But the NOK flag is not set, because the floating point value is not accurate. But Cpanel::JSON::XS switches to encoding the floating point value. JSON::XS also suffers from this issue.

**Comment by rurban** [comment]
I have to check  why JSON::XS choose that logic

**Comment by 17dec** [comment]
To add some context: JSON::XS only checks for the pNOK and pIOK flags and gives priority to the former. The reason for that priority makes sense, consider:

```perl
my $l = 1.5;
Dump($l);
int $l;
Dump($l);

__END__
SV = PVNV(0x562d601cd400) at 0x562d601fc580
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  IV = 412345678901234567
  NV = 1.5
  PV = 0
SV = PVNV(0x562d601cd400) at 0x562d601fc580
  REFCNT = 1
  FLAGS = (NOK,pIOK,pNOK)
  IV = 1
  NV = 1.5
  PV = 0
```
If the pIOK flag was given priority, encoding to JSON would throw away the fractional part.

The question remains if there was a reason for not considering the main NOK and IOK flags.

**Comment by rurban** [comment]
You are right. The fix is at line 2027: SvNOKp(sv) should be SvNOK(sv). The pNOK flag     
means "an NV slot exists," while NOK means "the NV value is currently valid/accurate." When a scalar  is upgraded via float operation but the NV is imprecise (IOK still set, NOK not set), we should use the accurate integer, not the imprecise float.

Need to add test for IOK+pNOK precision preservation
